File: internal/gitkit/ts/bin/tdd-classify.ts
// tdd-classify.ts — the deterministic, path-based half of the two-gate TDD classifier.
// Given a finding's file path, return the hard-skip category (always direct-fix, no test)
// or null (→ apply the behavioral gates: is it a defect? is it unit/intg-reproducible?).
// Run: node tdd-classify.ts <path>  → prints the category or "null". Erasable TS, zero deps.

export type SkipCategory = "migration" | "deps" | "ci" | "iac" | "generated" | "docs" | null;

// Ordered most-specific → least: ci before generic yaml/iac; migration before everything.
export function skipCategory(path: string): SkipCategory {
  const p = path.toLowerCase();

  if (/(^|\/)(migrations?|__migrations__)\//.test(p) || /(^|\/)migration\//.test(p)) return "migration";

  if (/(^|\/)(package-lock\.json|pnpm-lock\.yaml|yarn\.lock|bun\.lockb?|composer\.lock|gemfile\.lock|cargo\.lock|go\.sum|poetry\.lock)$/.test(p)) return "deps";

  if (/(^|\/)\.github\/workflows\//.test(p) || /(^|\/)\.gitlab-ci\.ya?ml$/.test(p) || /(^|\/)\.(circleci|buildkite)\//.test(p) || /(^|\/)azure-pipelines\.ya?ml$/.test(p)) return "ci";

  if (/(^|\/)dockerfile(\.[a-z0-9]+)?$/.test(p) || /docker-compose[^/]*\.ya?ml$/.test(p) || /\.(tf|tfvars)$/.test(p) || /(^|\/)(nginx|k8s|kubernetes|helm|terraform|ansible)\//.test(p) || /(^|\/)\.env(\.[a-z0-9]+)?$/.test(p)) return "iac";

  if (/\.(generated|gen)\.[a-z]+$/.test(p) || /\.d\.ts$/.test(p) || /(^|\/)(dist|build|out|\.next|coverage|__generated__|generated)\//.test(p)) return "generated";

  if (/\.(md|mdx|rst|txt)$/.test(p) || /(^|\/)docs?\//.test(p)) return "docs";

  return null;
}

async function main(): Promise<void> {
  const path = process.argv.slice(2).find((a) => !a.startsWith("--"));
  if (!path) throw new Error("usage: tdd-classify.ts <path>");
  process.stdout.write(`${skipCategory(path) ?? "null"}\n`);
}

if (import.meta.main) {
  main().catch((e) => {
    process.stderr.write(`error: ${(e as Error).message}\n`);
    process.exit(1);
  });
}

File: internal/gitkit/tddclassify.go
package gitkit

import (
	"errors"
	"fmt"
	"io"
	"regexp"
	"strings"
)

// tdd-classify — the deterministic, path-based half of the two-gate TDD
// classifier. A finding's file path maps to a hard-skip category (always
// direct-fix, no test) or "null" (→ apply the behavioral gates).

var skipRules = []struct {
	category string
	patterns []*regexp.Regexp
}{
	// Ordered most-specific → least: ci before generic yaml/iac; migration first.
	{"migration", compile(`(^|/)(migrations?|__migrations__)/`, `(^|/)migration/`)},
	{"deps", compile(`(^|/)(package-lock\.json|pnpm-lock\.yaml|yarn\.lock|bun\.lockb?|composer\.lock|gemfile\.lock|cargo\.lock|go\.sum|poetry\.lock)$`)},
	{"ci", compile(`(^|/)\.github/workflows/`, `(^|/)\.gitlab-ci\.ya?ml$`, `(^|/)\.(circleci|buildkite)/`, `(^|/)azure-pipelines\.ya?ml$`)},
	{"iac", compile(`(^|/)dockerfile(\.[a-z0-9]+)?$`, `docker-compose[^/]*\.ya?ml$`, `\.(tf|tfvars)$`, `(^|/)(nginx|k8s|kubernetes|helm|terraform|ansible)/`, `(^|/)\.env(\.[a-z0-9]+)?$`)},
	{"generated", compile(`\.(generated|gen)\.[a-z]+$`, `\.d\.ts$`, `(^|/)(dist|build|out|\.next|coverage|__generated__|generated)/`)},
	{"docs", compile(`\.(md|mdx|rst|txt)$`, `(^|/)docs?/`)},
}

func compile(patterns ...string) []*regexp.Regexp {
	out := make([]*regexp.Regexp, len(patterns))
	for i, p := range patterns {
		out[i] = regexp.MustCompile(p)
	}
	return out
}

// SkipCategory returns the hard-skip category for a path, or "" when the
// behavioral gates apply.
func SkipCategory(path string) string {
	p := strings.ToLower(path)
	for _, rule := range skipRules {
		for _, re := range rule.patterns {
			if re.MatchString(p) {
				return rule.category
			}
		}
	}
	return ""
}

// tddClassifyArgs is tdd-classify's argv: ONE path. A second path used to be
// dropped unclassified; --json is accepted (gitkit's own --json lands in a
// verb's argv) and changes nothing — the answer is one bare word.
var tddClassifyArgs = verbArgs{
	bools:       []string{"json"},
	positionals: 1,
	usage:       "usage: vybava gitkit tdd-classify <path>",
}

func runTDDClassify(args []string, stdout, stderr io.Writer) int {
	_, pos, err := tddClassifyArgs.parse("tdd-classify", args)
	if err != nil {
		return fail(stderr, err)
	}
	if len(pos) == 0 || pos[0] == "" {
		return fail(stderr, errors.New(tddClassifyArgs.usage))
	}
	path := pos[0]
	category := SkipCategory(path)
	if category == "" {
		category = "null"
	}
	fmt.Fprintln(stdout, category)
	return 0
}
